lib/proxy/peer/quicserver.go

```diff
@@ -78,31 +91,335 @@ func (c *QUICServerConfig) checkAndSetDefaults() error {
+	tlsConfig.ClientAuth = tls.RequireAndVerifyClientCert
+	tlsConfig.MinVersion = tls.VersionTLS13
+
+	getClientCAs := cfg.GetClientCAs
+	tlsConfig.GetConfigForClient = func(chi *tls.ClientHelloInfo) (*tls.Config, error) {
+		clientCAs, err := getClientCAs(chi)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		utils.RefreshTLSConfigTickets(tlsConfig)
+		c := tlsConfig.Clone()
+		c.ClientCAs = clientCAs
+		return c, nil
+	}
+
+	quicConfig := &quic.Config{
+		MaxStreamReceiveWindow:     quicMaxReceiveWindow,
+		MaxConnectionReceiveWindow: quicMaxReceiveWindow,
+
+		MaxIncomingStreams:    1 << 60,
+		MaxIncomingUniStreams: -1,
+
+		MaxIdleTimeout:  quicMaxIdleTimeout,
+		KeepAlivePeriod: quicKeepAlivePeriod,
+
+		Allow0RTT: true,
+	}
+
+	runCtx, runCancel := context.WithCancel(context.Background())
+	serveCtx, serveCancel := context.WithCancel(runCtx)
+
+	return &QUICServer{
+		log:           cfg.Log,
+		clusterDialer: cfg.ClusterDialer,
+		tlsConfig:     tlsConfig,
+		quicConfig:    quicConfig,
+
+		runCtx:      runCtx,
+		runCancel:   runCancel,
+		serveCtx:    serveCtx,
+		serveCancel: serveCancel,
+	}, nil
 }
 
 // Serve opens a listener and serves incoming connection. Returns after calling
 // Close or Shutdown.
 func (s *QUICServer) Serve(t *quic.Transport) error {
-	panic("QUIC proxy peering is not implemented")
+	s.mu.Lock()
+	if s.closed {
+		s.mu.Unlock()
+		return trace.Wrap(quic.ErrServerClosed)
+	}
+	s.wg.Add(1)
+	defer s.wg.Done()
+	s.mu.Unlock()
+
+	lis, err := t.ListenEarly(s.tlsConfig, s.quicConfig)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer lis.Close()
+	defer context.AfterFunc(s.serveCtx, func() { _ = lis.Close() })()
+
+	for {
+		// the listener will be closed when serveCtx is done, but Accept will
+		// return any queued connection before erroring out with a
+		// [quic.ErrServerClosed]
+		c, err := lis.Accept(context.Background())
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		s.wg.Add(1)
+		go s.handleConn(c)
+	}
+}
+
+func (s *QUICServer) handleConn(c quic.EarlyConnection) {
+	defer s.wg.Done()
+
+	log := s.log.With(
+		"remote_addr", c.RemoteAddr().String(),
+		"internal_id", uuid.NewString(),
+	)
+	state := c.ConnectionState()
+	log.InfoContext(s.serveCtx,
+		"Handling new peer connection.",
+		"gso", state.GSO,
+		"used_0rtt", state.Used0RTT,
+	)
+	defer func() {
+		log.DebugContext(c.Context(),
+			"Peer connection closed.",
+			"error", context.Cause(c.Context()),
+		)
+	}()
+
+	defer c.CloseWithError(0, "")
+	defer context.AfterFunc(s.runCtx, func() { _ = c.CloseWithError(0, "") })()
+
+	for {
+		// TODO(espadolini): stop accepting new streams once s.serveCtx is
+		// canceled, once quic-go gains the ability to change the amount of
+		// available streams during a connection (so we can set it to 0)
+		st, err := c.AcceptStream(context.Background())
+		if err != nil {
+			log.DebugContext(c.Context(), "Got an error accepting a stream.", "error", err)
```

Comment:
As a general comment, I'd rather this function returned an error and the caller made the choice to swallow it.

Reply:
The only possible error is caused by the connection getting closed - which is also why the log line is at debug level. I'm not convinced that moving the error logging one layer above will do much for the clarity of the code, at least while this is the only exit point for the function.

Reply:
Moving the logging up makes this behave like a regular erroring function, which is already a valuable readability improvement IMO.

Reply:
The logging happens before the defers, logging after returning would mean that the log line is related to an error that happened potentially much earlier, seeing as now we're waiting for the per-connection waitgroup to end.

Reply:
That is not typically an issue - a function errors, runs defers, returns the error and then the error gets logged by the caller. I won't push further but I do think a regular erroring func tends to be simpler to follow than the unusual over-swallowing of errors.